File: apps/idenplane/packages/idenplane-ios/Sources/Idenplane/BiometricAuth.swift
import Foundation
import LocalAuthentication

/// Wrapper around LocalAuthentication for Face ID / Touch ID gating of sensitive operations.
///
/// Use `BiometricAuth` to require biometric (or device passcode fallback) confirmation
/// before exposing tokens to the app. The prompt is shown each time `authenticate()` is called.
public final class BiometricAuth: @unchecked Sendable {

    // MARK: - Biometric type

    /// The biometric modality available on the current device.
    public enum BiometryType: Sendable {
        case faceID
        case touchID
        case opticID
        case none
    }

    // MARK: - Policy

    /// Whether to fall back to the device passcode when biometrics fail.
    public var allowPasscodeFallback: Bool

    public init(allowPasscodeFallback: Bool = true) {
        self.allowPasscodeFallback = allowPasscodeFallback
    }

    // MARK: - Capabilities

    /// Returns the biometric type available on this device, or `.none` if unavailable.
    public var biometryType: BiometryType {
        let context = LAContext()
        var error: NSError?
        guard context.canEvaluatePolicy(.deviceOwnerAuthenticationWithBiometrics, error: &error)
        else { return .none }

        switch context.biometryType {
        case .faceID:   return .faceID
        case .touchID:  return .touchID
        case .opticID:  return .opticID
        default:        return .none
        }
    }

    /// Returns `true` if the device supports any biometric modality.
    public var isBiometricAvailable: Bool {
        biometryType != .none
    }

    // MARK: - Authentication

    /// Prompt the user to authenticate via biometrics (or passcode if enabled).
    ///
    /// - Parameter reason: A human-readable string shown in the system prompt.
    /// - Throws: `IdenplaneError.biometricAuthFailed` if authentication is denied or unavailable.
    public func authenticate(reason: String = "Authenticate to access your account") async throws {
        let context = LAContext()
        context.localizedCancelTitle = "Cancel"

        let policy: LAPolicy = allowPasscodeFallback
            ? .deviceOwnerAuthentication
            : .deviceOwnerAuthenticationWithBiometrics

        var canEvalError: NSError?
        guard context.canEvaluatePolicy(policy, error: &canEvalError) else {
            let message = canEvalError?.localizedDescription ?? "Biometrics not available"
            throw IdenplaneError.biometricAuthFailed(message)
        }

        do {
            let success = try await context.evaluatePolicy(policy, localizedReason: reason)
            guard success else {
                throw IdenplaneError.biometricAuthFailed("Authentication was not successful")
            }
        } catch let laError as LAError {
            throw IdenplaneError.biometricAuthFailed(laError.localizedDescription)
        }
    }
}

File: apps/idenplane/packages/idenplane-android/src/main/kotlin/com/idenplane/sdk/BiometricAuth.kt
package com.idenplane.sdk

import android.content.Context
import androidx.biometric.BiometricManager
import androidx.biometric.BiometricManager.Authenticators
import androidx.biometric.BiometricPrompt
import androidx.core.content.ContextCompat
import androidx.fragment.app.FragmentActivity
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException
import kotlinx.coroutines.suspendCancellableCoroutine

/**
 * Wrapper around [BiometricPrompt] for requiring biometric or device-credential
 * authentication before granting access to sensitive token data.
 *
 * ## Usage
 * ```kotlin
 * val biometric = BiometricAuth(activity)
 * biometric.authenticate(title = "Verify your identity")
 * // Execution continues here only after successful authentication
 * val token = client.getAccessToken()
 * ```
 */
class BiometricAuth internal constructor(
    private val activity: FragmentActivity,
    private val biometricManager: BiometricManager,
) {

    /**
     * Public constructor — builds the real [BiometricManager] for [activity]. The two-arg
     * primary constructor above is `internal` purely so tests can inject a mocked
     * [BiometricManager], since there's no Robolectric shadow for it to drive real
     * hardware/enrollment states from a JVM unit test.
     */
    constructor(activity: FragmentActivity) : this(activity, BiometricManager.from(activity))

    // -----------------------------------------------------------------------
    // Availability
    // -----------------------------------------------------------------------

    /**
     * Returns `true` if the device supports biometric authentication AND the
     * user has enrolled at least one biometric credential.
     */
    val isBiometricAvailable: Boolean
        get() = biometricManager.canAuthenticate(
            Authenticators.BIOMETRIC_STRONG or Authenticators.BIOMETRIC_WEAK
        ) == BiometricManager.BIOMETRIC_SUCCESS

    /**
     * Returns a human-readable description of the biometric type available,
     * or `null` if biometrics are unavailable.
     */
    fun getBiometricType(context: Context): String? {
        val manager = BiometricManager.from(context)
        return describeBiometricType(
            strongResult = manager.canAuthenticate(Authenticators.BIOMETRIC_STRONG),
            weakResult = manager.canAuthenticate(Authenticators.BIOMETRIC_WEAK),
        )
    }

    companion object {
        /** Extracted so this mapping is unit-testable without needing a real BiometricManager. */
        internal fun describeBiometricType(strongResult: Int, weakResult: Int): String? =
            when (strongResult) {
                BiometricManager.BIOMETRIC_SUCCESS -> "Strong biometrics (fingerprint / face / iris)"
                else -> when (weakResult) {
                    BiometricManager.BIOMETRIC_SUCCESS -> "Weak biometrics"
                    else -> null
                }
            }
    }

    // -----------------------------------------------------------------------
    // Authentication
    // -----------------------------------------------------------------------

    /**
     * Show a biometric prompt and suspend until the user successfully authenticates.
     *
     * Falls back to device credentials (PIN / pattern / password) if biometrics fail
     * and [allowDeviceCredential] is `true`.
     *
     * @param title              Title shown in the biometric dialog.
     * @param subtitle           Optional subtitle shown in the dialog.
     * @param description        Optional description text.
     * @param negativeButtonText Text for the negative/cancel button (only shown
     *                           when [allowDeviceCredential] is `false`).
     * @param allowDeviceCredential Whether to fall back to PIN/pattern/password.
     *
     * @throws [IdenplaneException.BiometricAuthFailed] if authentication fails or is cancelled.
     */
    suspend fun authenticate(
        title: String = "Authenticate",
        subtitle: String? = null,
        description: String? = null,
        negativeButtonText: String = "Cancel",
        allowDeviceCredential: Boolean = true,
    ): Unit = suspendCancellableCoroutine { continuation ->

        val executor = ContextCompat.getMainExecutor(activity)

        val callback = object : BiometricPrompt.AuthenticationCallback() {
            override fun onAuthenticationSucceeded(result: BiometricPrompt.AuthenticationResult) {
                if (continuation.isActive) continuation.resume(Unit)
            }

            override fun onAuthenticationError(errorCode: Int, errString: CharSequence) {
                if (continuation.isActive) {
                    continuation.resumeWithException(
                        IdenplaneException.BiometricAuthFailed(errString.toString())
                    )
                }
            }

            override fun onAuthenticationFailed() {
                // Called when a biometric is presented but not recognized — the
                // system handles retry automatically; we only fail on error/cancel.
            }
        }

        val prompt = BiometricPrompt(activity, executor, callback)

        val info = BiometricPrompt.PromptInfo.Builder()
            .setTitle(title)
            .apply { subtitle?.let { setSubtitle(it) } }
            .apply { description?.let { setDescription(it) } }
            .apply {
                if (allowDeviceCredential) {
                    setAllowedAuthenticators(
                        Authenticators.BIOMETRIC_STRONG
                                or Authenticators.BIOMETRIC_WEAK
                                or Authenticators.DEVICE_CREDENTIAL
                    )
                } else {
                    setAllowedAuthenticators(
                        Authenticators.BIOMETRIC_STRONG or Authenticators.BIOMETRIC_WEAK
                    )
                    setNegativeButtonText(negativeButtonText)
                }
            }
            .build()

        continuation.invokeOnCancellation { prompt.cancelAuthentication() }
        prompt.authenticate(info)
    }
}
